Performance and memory improvements

bluelhf opened the pull request:
This PR adds small changes that improve the performance of Starlight, and synchronises the Starlight repository with PR [#8377 in PaperMC/Paper](https://github.com/PaperMC/Paper/pull/8377).

||Old|New|Change
|---|---|---|---
|Memory allocations|![The memory allocation call tree for the previous implementation](https://user-images.githubusercontent.com/52505120/191599753-347adaa1-6c8e-478b-873b-736307392385.png)|![The memory allocation call tree for the new implementation](https://user-images.githubusercontent.com/52505120/191599920-6ed620ea-f94e-41ea-a593-edb0af1b3b87.png)|**-23.1&nbsp;%**
|CPU samples|![The CPU sample call tree for the previous implementation](https://user-images.githubusercontent.com/52505120/191600298-ae5ca9c8-ce24-43cc-8ec5-6e1ab4b3d662.png)|![The CPU sample call tree for the new implementation](https://user-images.githubusercontent.com/52505120/191600502-76d4b9af-36ca-44ab-b98e-922c55a4604c.png)|**-35.4&nbsp;%**

<sup>Lower is better in both allocated bytes and CPU samples.</sup>

<sub>Copied from [Paper #8377](https://github.com/PaperMC/Paper/pull/8377)</sub>
This is done by swapping the CompletableFuture `thenCombine(...)` with a more efficient `thenCompose(...)` (we don't care about the result of the other CompletableFuture, as it does not have one). Additionally, a `HashSet` field in `StarlightInterface.LightQueue.ChunkTasks` was converted to an `ObjectOpenHashSet`. I did not measure the performance impact of this alone, but I would be surprised if the HashMap-backed java.util HashSet was faster than a 'real' implementation.

bluelhf committed:
perf(interface): use ObjectOpenHashSet instead of HashSet

sync for https://github.com/PaperMC/Paper/pull/8377

NoahvdAa commented:
The future call can also be updated for the mod, see this mixin:

https://github.com/PaperMC/Starlight/blob/a920f5a4e18ba126834b8136a159baa91344cc6e/src/main/java/ca/spottedleaf/starlight/mixin/common/lightengine/ThreadedLevelLightEngineMixin.java#L100-L102

bluelhf commented:
> The future call can also be updated for the mod, see this mixin:
> 
> https://github.com/PaperMC/Starlight/blob/a920f5a4e18ba126834b8136a159baa91344cc6e/src/main/java/ca/spottedleaf/starlight/mixin/common/lightengine/ThreadedLevelLightEngineMixin.java#L100-L102

Ah, I didn't notice that! I'll update it tomorrow. Thanks :)

bluelhf committed:
perf(lighting): use thenCompose(...) instead of thenCombine(...)

make the ThreadedLevelLightEngine compose the update future into neighbours' chunkToSave instead of combining them and discarding the void result, which is faster and saves memory

Spottedleaf commented on src/main/java/ca/spottedleaf/starlight/mixin/common/lightengine/ThreadedLevelLightEngineMixin.java:
```diff
@@ -97,9 +97,7 @@ private void queueTaskForSection(final int chunkX, final int chunkY, final int c
-                    neighbour.chunkToSave = neighbour.chunkToSave.thenCombine(updateFuture, (final ChunkAccess curr, final Void ignore) -> {
-                        return curr;
-                    });
+                    neighbour.chunkToSave = updateFuture.thenCompose(v -> neighbour.chunkToSave);
```
Looking over the code now, you could actually just remove this logic - the ticket should keep the chunks from unloading.

bluelhf replied:
Alright, I'll remove the logic ASAP :)

Still, I wonder what the original rationale for this code was — under what conditions does the server consider a world safe to save? Considering this code applies to neighbours, would it be necessary to add tickets for the neighbours as well?

Spottedleaf replied:
The chunk cannot unload until the chunkToSave is completed, so the idea was to stop the unload until the light task was done. But the ticket level should ensure that the chunk shouldn't unload, and the ticket cannot be removed until at least the light task is done.
Since the ticket level is 33, the neighbours will have ticket level 34. Chunks only unload once they exceed somewhere around 46, I don't remember off the top of my head, but given the chunk and its neighbours are above the threshold they wont be unloaded.

bluelhf replied:
Ah, thanks for the explanation!

bluelhf replied:
This should be resolved with 334500ca6e9c522924104586cac2e563430ec7ce :)

Spottedleaf reviewed (changes requested):
See comments

bluelhf committed:
refactor: bye bye neighbour composition
